### PR TITLE
User accounts

### PR DESCRIPTION
**Pull request by opadron**
Adds frontend UI components and logic for registering new accounts and logging into registered accounts.

Also significantly reorganizes and simplifies the frontend code base.  See commit messages for more details.

### CODE REVIEW

**Comment by opadron** [comment]
@osumo/developers any ideas?  Looking around, I've found nodejs/node-v0.x-archive#5504, but not much else.

**Comment by jeffbaumes** [comment]
To me it looks like it may have something to do with [routers](https://github.com/ReactTraining/react-router/pull/222) or other object which have many [listeners](https://pouchdb.com/errors.html#event_emitter_limit). It could also be a true memory leak. Could your router object have many (more than 10) listeners attached to it? The first link is a PR that makes the limit infinite. Seems a bit wrong to do that - perhaps we should see if increasing to 20 helps?

**Comment by opadron** [comment]
@osumo/developers can someone QA this?

File: web-external/src/components/dialog/login/footer.jsx
Comment (waxlamp) [inline web-external/src/components/dialog/login/footer.jsx]: s/Login/Log in/

Comment (waxlamp) [reply web-external/src/components/dialog/login/footer.jsx]: Get rid of the `NBSP` (looks weird on the button).

File: web-external/src/components/dialog/register/footer.jsx
Comment (waxlamp) [inline web-external/src/components/dialog/register/footer.jsx]: This makes `Register` appear in italics. Follow the example of `login/footer.jsx` and put a slash before the `i` tag's closing bracket, and remove the `</i>`.

**Review by waxlamp — CHANGES_REQUESTED** [review]
components/header/user.jsx, lines 45, 51, and 58 have the same "italic icon" bug as documented in the line notes for the register dialog.

**Comment by opadron** [comment]
I'm trying to merge this with the latest in master after landing #30.  It looks like I'm running into an issue I don't fully understand: it seems like every time I click on a file in the new file-selection widget, I end up getting redirected to the main page.

**Comment by opadron** [comment]
I've managed to figure out a lot of the kinks and get this work well integrated with #30.  I think it's working really well!  The only quirk I still notice is that the page scrolls to the bottom whenever you finish using the file-selector component.

@ronichoudhury please take a look at these latest commits, 9d4f9fd in particular.  I've summarized the majority of my latest contributions there.

**Comment by waxlamp** [comment]
I am still seeing italic text alongside icons, etc. (look, for instance, inside the Girder "username" dropdown menu at the right of the navbar).

**Comment by waxlamp** [comment]
I am having some trouble understanding the latest changes - but it's mainly a matter of my lack of background/depth with React.

My intuition is still telling me that something as simple as adding a file selector widget shouldn't cause such a dispersed changeset. I will take a closer look again in a bit.

**Comment by waxlamp** [comment]
Another comment - I noticed that the file selector doesn't _quite_ work the way it was before. Now, when I activate it, it starts off in my home folder, and I have no way to get to the collection containing the OSUMO input stuff. Is this by design?

**Comment by opadron** [comment]
> I am still seeing italic text alongside icons, etc. (look, for instance, inside the Girder "username" dropdown menu at the right of the navbar).

Sorry, I did all this integration work and completely forgot to address your original points.  I'll add the fixes, soon.

> My intuition is still telling me that something as simple as adding a file selector widget shouldn't cause such a dispersed changeset.

I think we should be careful to distinguish between dispersion over actual files and lines of code, and conceptual dispersion (both being present in these changesets).  The first we could always change by reorganizing how the code is laid out, but I would argue the latter is a direct consequence of the core tenets of React + Redux development, and we won't be able to do anything about it short of switching to something else.

Even for the simplest widgets, you're more likely than not to have to write code for the widget(s) themselves; code for container widget(s) to tie their props to the redux state tree; code that adds the new parts of the state tree that will be used in evaluating said props, including their reducers; code that implements the new actions that the widget(s) will trigger in their callbacks, and any new routes that would be mapped to these new actions.  We could move code around to keep changes limited to fewer files and lines of code at a time, but conceptually, there's a much more fundamental dispersion in React + Redux development that can't be helped.  Personally, I found that once I took the time to understand these patterns, it was much easier for me to go through these seemingly dispersed change sets and quickly make sense of them.

It is my sincere hope that you and @manthey would come to agree, but if you think the development style is really going to continue making things difficult for you, we should take a step back and reevaluate, while we still can.  Better for us to share an MO that we can all work with.

**Comment by opadron** [comment]
@ronichoudhury hopefully, I got all my italics slip ups.

Regarding the file browsing, I did tweak the behavior so that it always starts in your home space.  For my local instance, the data is under an `osumopublicuser` user's space, but if you have your data in a collection, then you are absolutely right that navigating to that is not possible in the UI (I think).

The machinery is in place to handle that navigation programmatically:

``` javascript
actions.setFileNavigation('collection', collectionId);
```

but as far as I can tell, the HierarchyView doesn't have the UI components needed to browse from collection to collection, or from one user's space to another.  It would be nice if there were a way to navigate to some "root" view, where users can see all the collections and users that they have permission to view.

I'd love to know what your thoughts are.

**Comment by waxlamp** [comment]
I uploaded a file to my user space so I could test, and everything seems to be working.

To make this approach general, we might want to have a script similar to https://github.com/Kitware/candela/blob/master/app/resonantlab/scripts/populateGirder.py to upload some data into some account's home folder.

I am not sure why the page jumps like that after coming back from the file selector, but the problem is not unique to that modal. If you do something like bring up the login dialog and then dismiss it, you will get the same page jumping effect. i.e., this seems to be connected to Girder modals. It doesn't happen on data.kitware.com, though, so I'm not sure what that means for OSUMO.

**Comment by manthey** [comment]
Should we be using the Browser widget rather than the Hierarchy widget?  This would let someone switch roots (different collections or users).

**Comment by waxlamp** [comment]
That sounds a lot like what we should be using :). Can that be "modalized"
the same way we have done with HierarchyWidget?

**Comment by opadron** [comment]
See [this](https://gist.github.com/opadron/8cd3f190eb2c77bf2017e9e1790ba363) patch (based on c2eb4c2) for my attempt at using multiple root divs to address the scrolling issue.  It looks like hoisting the dialog components to their own React component tree doesn't change the scrolling behavior.

I've checked all the computed values for the styling for pre-patch, post-patch, and compared it all to Girder's own styling and everything matches across all three cases, so I don't think it's a styling issue.

**Comment by opadron** [comment]
@manthey Check at the latest (c652548) changes.  I tried different approaches to using the BrowserWidget.  The solution I ended up going with was to just add the root selector piece.

Besides working pretty well, I'm also learning a lot about working Backbone into a React (+ Redux) app, which should help clear the way for more integration with girder core components.

Still not sure what to do about the scroll-jumping quirk.

**Comment by waxlamp** [comment]
We'll have to look a bit deeper into that "jumping" issue. I suspect it still has to do with the nesting issue @manthey identified. Maybe we can use a `position` style directive on the overlay CSS to influence how its drawn, etc., even when it's nested as it is with a React component. @manthey may have better ideas than me about that.

**Review by waxlamp — APPROVED** [review]

**Pull request merged**
